src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/HeadersProcessor.scala

```diff
@@ -257,34 +258,77 @@ trait HeadersProcessor extends ToDownloadProcessor with ScorexLogging with Score
 
   protected def heightIdsKey(height: Int): ByteArrayWrapper = ByteArrayWrapper(Algos.hash(Ints.toByteArray(height)))
 
+
+  private val EIP37VotingParameter: Byte = 8 // todo: should be 6 after testing , and 6 = 2100 in config
+
+  private val eip37Key = Blake2b256.hash("eip37 activation height")
+
+  private def storeEip37ActivationHeight(eip37ActivationHeight: Int) = {
+    historyStorage.insert(eip37Key, Ints.toByteArray(eip37ActivationHeight))
+  }
+
+  private def eip37VotedOn: Option[Int] = {
+    historyStorage.get(scorex.util.bytesToId(eip37Key)).map(Ints.fromByteArray)
+  }
+
   /**
     * Calculate difficulty for the next block
     *
     * @param parent - latest block
     * @return - difficulty for the next block
     */
   def requiredDifficultyAfter(parent: Header): Difficulty = {
-    if (parent.height == settings.chainSettings.voting.version2ActivationHeight || parent.height + 1 == settings.chainSettings.voting.version2ActivationHeight) {
-      // Set difficulty for version 2 activation height (where specific difficulty is needed due to PoW change)
-      settings.chainSettings.initialDifficultyVersion2
+    val parentHeight = parent.height
+
+    // todo: this EIP-37 activation checking code could be removed after activation
+    if (settings.chainSettings.isMainnet &&
+        parentHeight > 843776 &&
+        parentHeight <= 843776 + 8192 &&
+        parentHeight % 128 == 0 &&
+        eip37VotedOn.isEmpty) {
+      val chain = headerChainBack(512, parent, _ => false)
+      val eip37Activated = chain.headers.map(_.votes).map(_.contains(EIP37VotingParameter)).count(_ == true) >= 460
+      if (eip37Activated) {
+        storeEip37ActivationHeight(parentHeight + 1)
+      }
+    }
+
+    if (parentHeight > 843776 && parentHeight + 1 >= eip37VotedOn.getOrElse(Int.MaxValue)) {
+      // by eip37VotedOn definition could be on mainnet only
+      val epochLength = 128 // epoch length after EIP-37 activation
+      if (parentHeight % epochLength == 0) {
+        val heights = difficultyCalculator.previousHeadersRequiredForRecalculation(parentHeight + 1, epochLength)
+        // todo: if parent is on best chain, read headers directly, not via headerChainBack
+        val chain = headerChainBack(heights.max - heights.min + 1, parent, _ => false)
```

Comment:
What is the reasoning behind the `limit = heights.max - heights.min + 1` in `headerChainBack` ? Why to always increment it by `1` ?

Reply:
As header heights.max  counted, you need + 1 to catch heights.min. For example, if heights.max = 843776 , heights.max = 843668, you will end up  with 843669 without + 1